java/org/apache/catalina/realm/DataSourceRealm.java

```diff
@@ -539,6 +612,162 @@ private boolean isRoleStoreDefined() {
     }
 
 
+    /**
+     * Return the specified user's requested user attributes as a map.
+     * 
+     * @param dbConnection The database connection to be used
+     * @param username User name for which to return user attributes
+     * 
+     * @return a map containing the specified user's requested user attributes
+     */
+    protected Map<String, Object> getUserAttributesMap(Connection dbConnection, String username) {
+
+        String preparedAttributes = getUserAttributesStatement(dbConnection);
+        if (preparedAttributes == null || preparedAttributes == USER_ATTRIBUTES_NONE_REQUESTED) {
+            // The above reference comparison is intentional. USER_ATTRIBUTES_NONE_REQUESTED
+            // is a tag object (empty String) to distinguish between null (not yet
+            // initialized) and empty (no attributes requested).
+            // TODO Could as well be changed to `preparedAttributes.lenghth() = 0`
+
+            // Return null if no user attributes are requested (or if the statement was not
+            // yet built successfully)
+            return null;
+        }
+
+        try (PreparedStatement stmt = dbConnection.prepareStatement(preparedAttributes)) {
+            stmt.setString(1, username);
+
+            try (ResultSet rs = stmt.executeQuery()) {
+
+                if (rs.next()) {
+                    Map<String, Object> attrs = new LinkedHashMap<>();
```

Comment:
Why a linked hash map? Does it matter to the client to retain the order?

Reply:
Since it's a map, it's not essential. But `getAttributeNames()` returning the names in the order the attributes have been requested is simple an not expensive with a linked hash map.

Reply:
Do you know by chance whether the ServetRequest retains insertion/interation order? Maybe we could apply the same behavior.

Reply:
I don't know, but I guess the answer is no. AFAIK, using order preserving maps is a more recent thing (e. g. Google Chrome preserves insertion order of JavaScript Objects, which are HashMaps as well. That's also not required by the ECMA standards, but it's a neat feature, especially with debugging etc.)

> Maybe we could apply the same behavior.

Why not? But that's always a feature, users can't rely on. In fact, it's just _nice to have_, However, since `LinkedHashMap`is not much more costly than a classic `HashMap` (especially with only few entries), there is no pain with using it.